onnxruntime/core/framework/sequential_executor.cc

```diff
@@ -125,8 +126,71 @@ static Status ReleaseNodeMLValues(ExecutionFrame& frame,
 Status SequentialExecutor::Execute(const SessionState& session_state, const std::vector<int>& feed_mlvalue_idxs,
                                    const std::vector<OrtValue>& feeds, const std::vector<int>& fetch_mlvalue_idxs,
                                    std::vector<OrtValue>& fetches,
-                                   const std::unordered_map<size_t, CustomAllocator>& fetch_allocators,
-                                   const logging::Logger& logger) {
+                                   const std::unordered_map<size_t, IExecutor::CustomAllocator>& fetch_allocators,
+                                   const logging::Logger& logger,
+                                   int64_t& run_id) {
+  const SequentialExecutionPlan& seq_exec_plan = *session_state.GetExecutionPlan();
+  const auto& exec_plan_vec = seq_exec_plan.execution_plan;
+  const auto exec_plan_size = exec_plan_vec.size();
+  if (run_id == DEFAULT_RUN_ID) {
+    ExecutionFrame frame{feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches, fetch_allocators, session_state};
+    return Execute(session_state, feeds, fetch_mlvalue_idxs, fetches, logger, frame, 0, exec_plan_size);
+  } else {
+    // Partial graph execution frame management code.
+    ExecutionFrame* frame;
+    std::map<int64_t, std::unique_ptr<onnxruntime::ExecutionFrame>>::iterator it;
+    if (run_id == DEFAULT_PARTIAL_RUN_ID) {
+      auto p_frame = onnxruntime::make_unique<ExecutionFrame>(feed_mlvalue_idxs, feeds, fetch_mlvalue_idxs, fetches,
+                                                              fetch_allocators, session_state);
+
+      frame = p_frame.get();
+      std::lock_guard<OrtMutex> lock(session_state.graph_runs_lock_);
+      it = session_state.graph_runs_.insert(std::make_pair(session_state.graph_runs_counter_, std::move(p_frame))).first;
+      ORT_ENFORCE(run_id < INT64_MAX);
+      run_id = session_state.graph_runs_counter_;
+      // REVIEW(mzs): May be reuse counters and prevent a potential overflow?
+      session_state.graph_runs_counter_ += 1;
+    } else {
+      ORT_ENFORCE(run_id >= 0);
+
+      it = session_state.graph_runs_.find(run_id);
+
+      ORT_ENFORCE(it != session_state.graph_runs_.end());
+
+      (it->second)->UpdateFeedAndFetches(feed_mlvalue_idxs, feeds, const_cast<std::vector<int>&>(fetch_mlvalue_idxs), fetches);
```

Comment:
Yes, it would be cleaner. I'm not sure if we should be updating fetches after execute because sometimes the user passes the tensor for fetches and we use that and avoid an explicit copy in the end, in this scenario the node could be directly writing to that user supplied tensor and we would want the tensor table in execution frame to have up-to-date fetch tensor, moreover updating fetches after execution would also be a departure from the current behavior. I could be wrong but this is my understanding and will cross check again. I have broken this function into two.

---
In reply to: [598932463](https://github.com/microsoft/onnxruntime/pull/7077#discussion_r598932463) [](ancestors = 598932463)